app/code/core/Mage/Core/Model/Config.php

```diff
@@ -736,17 +801,16 @@ protected function _getDeclaredModuleFiles()
             $name = explode(DIRECTORY_SEPARATOR, $v);
             $name = substr($name[count($name) - 1], 0, -4);
 
-            if ($name == 'Mage_All') {
-                $collectModuleFiles['base'][] = $v;
-            } elseif (substr($name, 0, 5) == 'Mage_') {
-                $collectModuleFiles['mage'][] = $v;
+            if (array_key_exists($name, self::MAGE_MODULES)) {
+                $collectModuleFiles['mage'][self::MAGE_MODULES[$name]] = $v;
             } else {
                 $collectModuleFiles['custom'][] = $v;
             }
         }
 
+        ksort($collectModuleFiles['mage']);
```

Comment:
isn't the ksort changing the loading order of the MAGE_MODULES array? maybe I'm missing something

Reply:
Good observation. The ksort function performs ascending sorting by key and not by value. The modules will be loaded in alphabetical order. Mage_A, Mage_B, ...

Reply:
Ehm... no they don't (as you can also see in the screenshots). I double checked this. Please don't confuse `$collectModuleFiles['mage']` (just the mage part) with `$sortedConfig` (the final loading sort order).

But let's prove it:
Before ksort:
![grafik](https://user-images.githubusercontent.com/12640514/185759183-1eb06615-35be-4cc2-9e46-fbb80616988d.png)

After ksort:
![grafik](https://user-images.githubusercontent.com/12640514/185759206-027f7e9c-811d-4c6d-9769-42706b41e283.png)

And just to be sure `$sortedConfig` for the final result:
![grafik](https://user-images.githubusercontent.com/12640514/185759260-00c72788-e6f7-4edf-a2eb-dfe6593328b2.png)

Please excuse me if I misunderstood your hints. :)

Reply:
mm sorry then i didn't understand the code. why is the ksort needed?

Reply:
No problem, I'll try to tell you. All module XMLs are read in a foreach loop (alphabetical and all together from `app/etc/modules`. They're kind of presorted into 'mage' (all Mage modules) and 'custom' (all other modules). We're only talking about the mage part here. Because they come in alphabetical from the foreach they're of course alphabetical at the beginning. But with MAGE_MODULES they get their sort order on by another (as array key). But the array isn't sorted automatically afterwards. This is why the ksort is added after all modules are read. After the ksort all Mage modules are in correct order. And all other/custom modules are loaded after all Mage modules, because of the array_merge of the (sorted) 'mage' and (untouched) 'custom' part at the end.